File: LintCode-Python/jumpGame.py
# Given an array of non-negative integers, you are initially positioned at the
# first index of the array.

# Each element in the array represents your maximum jump length at that position.

# Determine if you are able to reach the last index.

# Have you met this question in a real interview? Yes
# Example
# A = [2,3,1,1,4], return true.

# A = [3,2,1,0,4], return false.

# Note
# This problem have two method which is Greedy and Dynamic Programming.

# The time complexity of Greedy method is O(n).

# The time complexity of Dynamic Programming method is O(n^2).

# We manually set the small data set to allow you pass the test in both ways.
# This is just to let you learn how to use this problem in dynamic programming
# ways. If you finish it in dynamic programming ways, you can try greedy method
# to make it accept again.


# greedy
def jumpGameI(A):
    if not A:
        return False
    n = len(A)
    farthest = A[0]
    for i in range(n):
        if i <= farthest and i + A[i] >= farthest:
            farthest = i + A[i]
    return farthest >= n - 1





# DP
def jumpGame(A):
    if not A:
        return False
    n = len(A)
    can = [False for i in range(n)]

    # init
    can[0] = True

    # func
    for i in range(1, n):
        for j in range(i):
            if can[j] and j + A[j] >= i:
                can[i] = True
                break

    # answer:
    return can[n - 1]


print jumpGame([2, 3, 1, 1, 4])
print jumpGame([3, 2, 1, 0, 4])

print jumpGameI([2, 3, 1, 1, 4])
print jumpGameI([3, 2, 1, 0, 4])

File: LeetCode_Java/src/JumpGame.java
/**
 * Given an array of non-negative integers, you are initially positioned at the first index of the array.
 * Each element in the array represents your maximum jump length at that position.
 * Determine if you are able to reach the last index.
 * For example:
 * A = [2,3,1,1,4], return true.
 * A = [3,2,1,0,4], return false.
 * 
 * Tags: Array, Greedy
 * @author wendi
 *
 */
public class JumpGame {
	
	/**
	 * Method2: DP --> Greedy: maxIndex = i + nums[i], where 0<=i<nums.length.
	 * If there is a false in the process, the result will be false.
	 * @param int[] nums
	 * @return boolean
	 * Time: O(n)
	 * Space: O(1) optimize dpO(n) to O(1)
	 */
	public boolean jumpGameI(int[] nums) {
		if (nums == null || nums.length == 0) return true;
        int maxIndex = 0;
        for (int i = 0; i < nums.length; i++) {
        	if (i > maxIndex) return false;
        	maxIndex = Math.max(i + nums[i], maxIndex);
        }
        return true;
	}
	
	
	/**
	 * Method1: DP(brute force)
	 * @param int[] nums
	 * @return boolean
	 * Time: O(n)
	 * Space: O(n)
	 */
	public boolean jumpGame(int[] nums) {
        if (nums == null || nums.length == 0) return false;
        if (nums.length == 1) return true;
        int n = nums.length;
        boolean[] dp = new boolean[n];
        dp[0] = true;
        for (int i = 0; i < n; i++) {
            if (!dp[i]) continue;
            for (int j = 1; j <= nums[i]; j++) {
                if (i + j < n) dp[i + j] = true;
                if (dp[n - 1]) return true;
            }
        }
        return false;
	}
		

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		JumpGame result = new JumpGame();
		System.out.println(result.jumpGame(new int[] {2,3,1,1,4}));
//		System.out.println(result.jumpGame(new int[] {3,2,1,0,4}));
		System.out.println(result.jumpGameI(new int[] {2,3,1,1,4}));
	}

}
